Non recursive add simplex

maximelucas opened the pull request:
- rewrite of  `add_simplices_from` so that is is non-recursive.
- updated tests

With the rewrite, we first add all input simplices, store all faces, and add the faces add the end. It used to add each input simplex with its faces, so this changes the order in which simplices are added. This new order solves #244 

Notes: 
- to avoid too much redundancy and slowness, we add the `set` of faces. This seems to sometimes randomly change the order of simplices.
- one test does not pass: `test_boundary_matrix`. I think it's cause by the change in order which causes simplices to have different IDs. Might need @Marconurisso to check on this one. 
- need to check if #85 is still a problem

nwlandry commented on xgi/classes/simplicialcomplex.py:
```diff
@@ -214,32 +218,52 @@ def add_simplex(self, members, id=None, **attr):
+        # add all subfaces
+        faces = self._subfaces(members)
+        faces = set(faces)  # get unique faces
+        for members in faces:
```
Is it a problem that `members` from the input is getting overwritten here?

maximelucas replied:
Probably not but I changed it to be sure, thanks

maximelucas commented on xgi/classes/simplicialcomplex.py:
```diff
@@ -189,8 +193,8 @@ def add_simplex(self, members, id=None, **attr):
         >>> S.edges.members() # doctest: +NORMALIZE_WHITESPACE
-        [frozenset({1, 2, 3}), frozenset({1, 2}),
-            frozenset({1, 3}), frozenset({2, 3})]
+        [frozenset({1, 2, 3}), frozenset({2, 3}),
```
From what I've seen, in this case yes. I noticed the order was not reliable with string nodes.. so I tested for set equality in those cases.

maximelucas replied:
But Github action might be proving me wrong right now, this order thing is all a bit mysterious still. For strings I noticed it would change only when restarting the python kernel for example.

maximelucas replied:
From Github actions it seems order is deterministic in 3.11 but random in 3.7... there are more tests that don't pass in 3.7.

maximelucas replied:
Update: not random in 3.7, deterministic. But consistently a different order than in 3.11.

maximelucas replied:
Done: checking for set equality now.

maximelucas commented:
- `test_boundary_matrix` now passes - thanks @Marconurisso 
- #85 is fixed
- put the warning tests back in: they were slow because of an infinite loop - now fixed
- versions 3.7 and 3.10 of python have a different "order" for sets. Added some +SKIP flags in doctests when they failed because of that.

nwlandry commented:
I did a few tests of my own, adding/removing simplices, etc. and everything works flawlessly! This is great work!!
